include/xtensor/xscalar.hpp

```diff
@@ -973,13 +974,21 @@ namespace xt
     template <bool is_const, class CT>
     inline void xscalar_stepper<is_const, CT>::to_begin() noexcept
     {
-        p_c = p_c->stepper_begin(p_c->shape()).p_c;
+        if (m_end)
+        {
+            p_c = p_c->stepper_begin(p_c->shape()).p_c;
+            m_end = false;
+        }
     }
 
     template <bool is_const, class CT>
     inline void xscalar_stepper<is_const, CT>::to_end(layout_type l) noexcept
     {
-        p_c = p_c->stepper_end(p_c->shape(), l).p_c;
+        if (!m_end)
+        {
+            p_c = p_c->stepper_end(p_c->shape(), l).p_c;
```

Comment:
similar here? or will it end up being the same code as the compiler can optimize this stuff?